BUILDING.md

```diff
@@ -0,0 +1,578 @@
+`build/builder/mkrelease.sh` on TeamCity agents.
+
+The `build/packer/` directory contains what is needed for building VM images for the TeamCity
+agents.
+
+The `build/node-run.sh` script is a hack that pipes NodeJS's stdout and stderr through `cat` which
+prevents the actual stdout and stderr streams from being infected with non-blocking I/O. The
+comments in the script fully explain the scenario. It's used in the Makefile for each node-related
+command.
+
+The `build/upload-coverage.sh` script generates code coverage reports of all packages in the `pkg`
+directory and uploads the coverage profile to codecov.io. It's run via `make upload-coverage`.
+
+The `bin/uptodate` program is built from `pkg/cmd/uptodate` and used to check if c-deps are
+up-to-date or need to be rebuilt.
+
+The `build/werror.sh` script promotes warnings produced by a command to fatal errors, like GCC's
+`-Werror` flag.
+
+The `bin/execgen` program is built from `pkg/sql/colexec/execgen/cmd/execgen` and used to generate
+templated code related to columnarized execution.
+
+A slew of additional binaries deposited in `bin/` can be built using `make`. See [Go binary
+dependency file generation](#go-binary-dependency-file-generation) below for the list.
+
+
+# Tool-Specific Details
+
+## Make
+
+The CockroachDB makefile is large and complex. Here we focus on summarizing and clarifying it.
+
+`make help` gives an overview of available targets, variables and usage examples. We do not
+duplicate that information here.
+
+There are many detailed and helpful comments throughout the Makefile. When information from these
+comments is included in this document, it is only in summary.
+
+This make file can:
+
+- run on GNU and BSD systems (many commands have flags set by variables)
+- be run with different versions of Make (not just GNU make)
+- cross-compile cockroach to darwin, linux-gnu, linux-musl, and windows
+
+
+### Approximate Sections of the Makefile
+
+Functionality is mostly split into relevant sections. These are the approximate line numbers for
+each section (they will surely change). Additonal details on each section are included below.
+
+- Initialization: 0 - 374
+- C deps: about 375 - 785
+- Cockroach: 801 - 965
+- Install/start cockroach: 966 - 976
+- Cockroach testing: 977 - 1075
+- Auxiliary commands: 1076 - 1159
+- .buildinfo: 1160 - 1183
+- Protobuf configuration: 1184 - 1307
+- UI: 1308 - 1414
+- SQL parser: 1415 - 1587
+- c-deps-fmt: 1588 - 1592
+- Cleaning: 1593 - 1638
+- Go binary dependency file generation: 1639 - 1732
+- Check CLI variables are all valid: 1733 - EOF
+
+
+#### Initialization Steps
+
+1. The `GO` and `xgo` (needed for `defs.mk`) variables are set.
+
+1. Some variables need to be computed using shell commands. The first step in the make file is to compute and cache a set of them in the `build/defs.mk` file so they don't get recomputed on every run of make. `build/defs.mk` is then included in `Makefile`.
```

Comment:
I'm also nervous about this kind of documentation. It's nearly as detailed as the Makefile itself. Who is going to read this and how is it going to help them? Most importantly, who is going to update it as the Makefile changes?

If this documentation was helpful for you to write, would it have been helpful to add as Makefile comments?